cpp/embeddings.cc

```diff
@@ -0,0 +1,42 @@
+#include <cstring>
+#include  <stdexcept>
+
+#include "embeddings.h"
+#include "finalfusion.h"
+
+Embeddings::Embeddings(std::string const &filename, bool const mmap) {
+  ff_embeddings embeddings;
+  if (mmap) {
+    embeddings = ff_mmap_embeddings(filename.c_str());
+  } else {
+    embeddings = ff_read_embeddings(filename.c_str());
+  }
+
+  if (embeddings == nullptr) {
+    throw std::runtime_error(std::string(ff_error()));
+  }
+  d_inner = embeddings;
+}
+
+Embeddings::~Embeddings() {
+  ff_free_embeddings(d_inner);
+}
+float *Embeddings::embeddingRaw(std::string const &word) {
+  return ff_embedding_lookup(d_inner, word.c_str());
+}
+
+std::vector<float> Embeddings::embedding(std::string const &word) {
```

Comment:
Worth considering: C++ 17 has an option type (`std::optional`).

Reply:
tensorflow is built in C++ 14 per default:

https://github.com/tensorflow/tensorflow/blob/17f1a61ad924ee8c259f420e7602f6d31cf8d734/.bazelrc#L115-L117

Would using C++ 17 features have implications on the usability with default compiled TF versions then? Although it looks like there's also a switch to compile with C++ 17 features.

Reply:
In general I had no problems compiling e.g. C++03 code with C++11, unless they used overlapping type names, etc. Just try to compile it with C++17 and see what happens.

I am fine with sticking with C++11 or C++14 as well.